File: src/scanner.js
"use strict";

const fs = require("fs");
const path = require("path");

// Directories to always skip
const IGNORE_DIRS = new Set([
  "node_modules", ".git", "__pycache__", ".venv", "venv", "env",
  ".next", ".nuxt", "dist", "build", ".turbo", ".cache",
  ".mypy_cache", ".pytest_cache", ".ruff_cache", "htmlcov",
  "coverage", ".tox", "eggs", "*.egg-info", ".terraform",
]);

// File extensions grouped by category
const CATEGORIES = {
  backend:    [".py", ".pyi"],
  frontend:   [".tsx", ".jsx", ".ts", ".js", ".vue", ".svelte"],
  config:     [".json", ".yaml", ".yml", ".toml", ".ini", ".cfg", ".env"],
  style:      [".css", ".scss", ".sass", ".less"],
  template:   [".html", ".ejs", ".hbs", ".jinja2", ".j2"],
  data:       [".sql", ".graphql", ".gql", ".prisma"],
  docs:       [".md", ".rst", ".txt"],
  docker:     ["Dockerfile", "docker-compose.yml", "docker-compose.yaml"],
  ci:         [".github", ".gitlab-ci.yml", "Jenkinsfile"],
  shell:      [".sh", ".bash", ".zsh", ".ps1"],
  extension:  [".crx"],
};

function getCategory(filePath) {
  const base = path.basename(filePath);
  const ext = path.extname(filePath).toLowerCase();

  if (base === "Dockerfile" || base.startsWith("docker-compose")) return "docker";
  if (base === "Jenkinsfile") return "ci";
  if (filePath.includes(".github")) return "ci";
  if (base === "manifest.json" && filePath.includes("extension")) return "extension";

  for (const [cat, exts] of Object.entries(CATEGORIES)) {
    if (exts.includes(ext)) return cat;
  }
  return "other";
}

function getLanguage(filePath) {
  const ext = path.extname(filePath).toLowerCase();
  const map = {
    ".py": "python", ".pyi": "python",
    ".js": "javascript", ".jsx": "javascript",
    ".ts": "typescript", ".tsx": "typescript",
    ".json": "json", ".yaml": "yaml", ".yml": "yaml",
    ".toml": "toml", ".md": "markdown",
    ".html": "html", ".css": "css", ".scss": "scss",
    ".sql": "sql", ".sh": "shell", ".bash": "shell",
    ".graphql": "graphql", ".prisma": "prisma",
    ".vue": "vue", ".svelte": "svelte",
    ".rs": "rust", ".go": "go", ".java": "java",
    ".rb": "ruby", ".php": "php", ".swift": "swift",
    ".kt": "kotlin", ".cs": "csharp", ".cpp": "cpp",
    ".c": "c", ".h": "c", ".hpp": "cpp",
  };
  return map[ext] || "text";
}

/**
 * Scan a project directory and build a file tree with metadata.
 * Returns { tree, files, stats }
 */
function scanProject(rootDir, opts = {}) {
  const maxDepth = opts.maxDepth || 12;
  const maxFiles = opts.maxFiles || 5000;
  const files = [];
  let fileCount = 0;

  function shouldIgnore(name) {
    if (name.startsWith(".") && name !== ".claude" && name !== ".github") return true;
    return IGNORE_DIRS.has(name);
  }

  function walkDir(dir, depth, relPath) {
    if (depth > maxDepth || fileCount >= maxFiles) return null;

    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return null;
    }

    const node = {
      name: path.basename(dir),
      path: relPath || ".",
      type: "directory",
      children: [],
    };

    // Sort: dirs first, then files, alphabetical within each
    entries.sort((a, b) => {
      if (a.isDirectory() && !b.isDirectory()) return -1;
      if (!a.isDirectory() && b.isDirectory()) return 1;
      return a.name.localeCompare(b.name);
    });

    for (const entry of entries) {
      const entryPath = path.join(dir, entry.name);
      const entryRelPath = relPath ? `${relPath}/${entry.name}` : entry.name;

      if (entry.isDirectory()) {
        if (shouldIgnore(entry.name)) continue;
        const child = walkDir(entryPath, depth + 1, entryRelPath);
        if (child && child.children.length > 0) {
          node.children.push(child);
        }
      } else if (entry.isFile()) {
        if (fileCount >= maxFiles) break;
        fileCount++;

        let stat;
        try {
          stat = fs.statSync(entryPath);
        } catch {
          continue;
        }

        const fileNode = {
          name: entry.name,
          path: entryRelPath,
          type: "file",
          size: stat.size,
          modified: stat.mtimeMs,
          category: getCategory(entryRelPath),
          language: getLanguage(entryRelPath),
        };

        node.children.push(fileNode);
        files.push(fileNode);
      }
    }

    return node;
  }

  const tree = walkDir(rootDir, 0, "");

  // Build stats
  const stats = {
    totalFiles: files.length,
    byCategory: {},
    byLanguage: {},
    totalSize: 0,
  };

  for (const f of files) {
    stats.byCategory[f.category] = (stats.byCategory[f.category] || 0) + 1;
    stats.byLanguage[f.language] = (stats.byLanguage[f.language] || 0) + 1;
    stats.totalSize += f.size;
  }

  return { tree, files, stats };
}

/**
 * Detect project components (backend, frontend, extension, etc.)
 */
function detectComponents(rootDir, tree) {
  const components = [];

  function findDir(node, name) {
    if (!node || !node.children) return null;
    for (const child of node.children) {
      if (child.type === "directory" && child.name === name) return child;
    }
    return null;
  }

  function hasFile(dirPath, fileName) {
    try { return fs.existsSync(path.join(rootDir, dirPath, fileName)); }
    catch { return false; }
  }

  // Detect backend — check multiple common layouts
  const backendDetected = (() => {
    // Monorepo: backend/ directory
    if (hasFile("backend", "pyproject.toml") || hasFile("backend", "requirements.txt")) {
      return { path: "backend" };
    }
    // Python src/ layout (sales-coach pattern)
    if (hasFile("src", "main.py") || hasFile("src", "__init__.py")) {
      return { path: "src" };
    }
    // Root-level Python project
    if (hasFile("", "pyproject.toml") || hasFile("", "requirements.txt")) {
      // Check if there's an app/ or src/ dir with Python files
      if (hasFile("app", "main.py") || hasFile("app", "__init__.py")) return { path: "app" };
      if (hasFile("src", "main.py")) return { path: "src" };
      return { path: "." };
    }
    return null;
  })();

  if (backendDetected) {
    // Detect specific framework
    let tech = "Python";
    try {
      const pyFiles = ["pyproject.toml", "requirements.txt"].map((f) =>
        [path.join(rootDir, f), path.join(rootDir, backendDetected.path, f)]
      ).flat();
      for (const fp of pyFiles) {
        if (fs.existsSync(fp)) {
          const txt = fs.readFileSync(fp, "utf8").toLowerCase();
          if (txt.includes("fastapi")) { tech = "Python / FastAPI"; break; }
          if (txt.includes("django")) { tech = "Python / Django"; break; }
          if (txt.includes("flask")) { tech = "Python / Flask"; break; }
        }
      }
    } catch { /* fallback */ }

    components.push({
      name: backendDetected.path === "src" ? "Source" : "Backend",
      type: "backend",
      path: backendDetected.path,
      tech,
      icon: "🐍",
    });
  }

  // Detect frontend / portal
  for (const dir of ["portal", "frontend", "web", "app", "client"]) {
    if (hasFile(dir, "package.json")) {
      const pkgPath = path.join(rootDir, dir, "package.json");
      try {
        const pkg = JSON.parse(fs.readFileSync(pkgPath, "utf8"));
        const deps = { ...pkg.dependencies, ...pkg.devDependencies };
        let tech = "Node.js";
        if (deps.next) tech = "Next.js";
        else if (deps.react) tech = "React";
        else if (deps.vue) tech = "Vue";
        else if (deps.svelte) tech = "Svelte";
        else if (deps["@angular/core"]) tech = "Angular";

        components.push({
          name: dir.charAt(0).toUpperCase() + dir.slice(1),
          type: "frontend",
          path: dir,
          tech,
          icon: "⚛️",
        });
      } catch { /* skip */ }
      break;
    }
  }

  // Detect Chrome extension
  for (const dir of ["chrome-extension", "extension", "ext"]) {
    if (hasFile(dir, "manifest.json")) {
      components.push({
        name: "Chrome Extension",
        type: "extension",
        path: dir,
        tech: "Chrome MV3",
        icon: "🧩",
      });
      break;
    }
  }

  // Detect .claude config
  if (hasFile(".claude", "settings.json") || hasFile("", "CLAUDE.md")) {
    components.push({
      name: "Claude Config",
      type: "config",
      path: ".claude",
      tech: "Claude Code",
      icon: "🤖",
    });
  }

  // Detect Docker
  if (hasFile("", "Dockerfile") || hasFile("", "docker-compose.yml") || hasFile("backend", "Dockerfile")) {
    components.push({
      name: "Docker",
      type: "docker",
      path: ".",
      tech: "Docker / Compose",
      icon: "🐳",
    });
  }

  // Detect CI
  if (hasFile(".github/workflows", "ci.yml") || hasFile(".github/workflows", "ci.yaml")) {
    components.push({
      name: "CI/CD",
      type: "ci",
      path: ".github/workflows",
      tech: "GitHub Actions",
      icon: "⚡",
    });
  }

  // Detect YAML config directory
  for (const dir of ["config", "conf", "settings"]) {
    try {
      const dirPath = path.join(rootDir, dir);
      if (fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory()) {
        const yamlFiles = fs.readdirSync(dirPath).filter((f) => f.endsWith(".yaml") || f.endsWith(".yml") || f.endsWith(".json"));
        if (yamlFiles.length > 0) {
          components.push({
            name: "Config",
            type: "config",
            path: dir,
            tech: `YAML/JSON (${yamlFiles.length} files)`,
            icon: "⚙️",
          });
          break;
        }
      }
    } catch { /* skip */ }
  }

  // Detect knowledge / data layer
  for (const dir of ["knowledge", "data", "models"]) {
    try {
      const dirPath = path.join(rootDir, dir);
      if (fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory()) {
        const files = fs.readdirSync(dirPath);
        if (files.length > 0) {
          components.push({
            name: dir.charAt(0).toUpperCase() + dir.slice(1),
            type: "database",
            path: dir,
            tech: "Data / Knowledge",
            icon: "🗄️",
          });
          break;
        }
      }
    } catch { /* skip */ }
  }

  return components;
}

module.exports = { scanProject, detectComponents, getCategory, getLanguage };

File: arch_viewer/scanner.py
"""
Static code scanner — builds architecture model WITHOUT AI.
This is the fast, deterministic baseline. The AI agent enriches it.
"""

from __future__ import annotations

import json
import os
import re
from pathlib import Path

from .models import (
    APIRoute,
    Architecture,
    Component,
    ComponentType,
    DataFlow,
    DataFlowDirection,
    DependencyInfo,
    FileInfo,
)

IGNORE_DIRS = {
    "node_modules", ".git", "__pycache__", ".venv", "venv", "env",
    ".next", ".nuxt", "dist", "build", ".turbo", ".cache",
    ".mypy_cache", ".pytest_cache", ".ruff_cache", "htmlcov",
    "coverage", ".tox", ".terraform", ".svn",
}

EXT_LANG = {
    ".py": "python", ".pyi": "python", ".js": "javascript", ".jsx": "javascript",
    ".ts": "typescript", ".tsx": "typescript", ".json": "json", ".yaml": "yaml",
    ".yml": "yaml", ".toml": "toml", ".md": "markdown", ".html": "html",
    ".css": "css", ".scss": "scss", ".sql": "sql", ".sh": "shell",
    ".bash": "shell", ".graphql": "graphql", ".prisma": "prisma",
    ".vue": "vue", ".svelte": "svelte", ".rs": "rust", ".go": "go",
    ".java": "java", ".rb": "ruby", ".php": "php", ".cs": "csharp",
    ".cpp": "cpp", ".c": "c", ".h": "c", ".hpp": "cpp", ".kt": "kotlin",
    ".swift": "swift", ".r": "r", ".jl": "julia",
}

EXT_CATEGORY = {
    ".py": "backend", ".pyi": "backend",
    ".tsx": "frontend", ".jsx": "frontend",
    ".ts": "frontend", ".js": "frontend",
    ".vue": "frontend", ".svelte": "frontend",
    ".json": "config", ".yaml": "config", ".yml": "config",
    ".toml": "config", ".ini": "config", ".cfg": "config",
    ".css": "style", ".scss": "style", ".sass": "style", ".less": "style",
    ".html": "template", ".ejs": "template", ".hbs": "template",
    ".sql": "data", ".graphql": "data", ".prisma": "data",
    ".md": "docs", ".rst": "docs", ".txt": "docs",
    ".sh": "shell", ".bash": "shell", ".ps1": "shell",
}


def scan_project(root: str | Path, max_files: int = 5000) -> Architecture:
    """Full static scan — produces an Architecture model ready for AI enrichment."""
    root = Path(root).resolve()
    project_name = root.name

    all_files: list[FileInfo] = []
    tree: dict = {}
    file_count = 0

    for dirpath, dirnames, filenames in os.walk(root):
        # Prune ignored dirs
        dirnames[:] = [
            d for d in dirnames
            if d not in IGNORE_DIRS and not (d.startswith(".") and d not in (".claude", ".github"))
        ]

        rel_dir = os.path.relpath(dirpath, root)
        if rel_dir == ".":
            rel_dir = ""

        for fname in sorted(filenames):
            if file_count >= max_files:
                break

            fpath = os.path.join(dirpath, fname)
            rel_path = os.path.join(rel_dir, fname).replace("\\", "/") if rel_dir else fname

            try:
                stat = os.stat(fpath)
            except OSError:
                continue

            ext = os.path.splitext(fname)[1].lower()
            lang = EXT_LANG.get(ext, "text")
            cat = EXT_CATEGORY.get(ext, "other")

            # Special cases
            if fname == "Dockerfile":
                cat = "docker"
            elif fname.startswith("docker-compose"):
                cat = "docker"
            elif ".github" in rel_path:
                cat = "ci"

            fi = FileInfo(
                path=rel_path,
                language=lang,
                category=cat,
                size=stat.st_size,
                modified=stat.st_mtime,
            )

            # Extract imports for Python files (lightweight)
            if lang == "python" and stat.st_size < 500_000:
                fi.imports = _extract_python_imports(fpath)

            all_files.append(fi)
            file_count += 1

            # Build nested tree dict
            _insert_tree(tree, rel_path.split("/"), fi)

    components = _detect_components(root, all_files)
    data_flows = _infer_data_flows(components, all_files)
    deps = _collect_dependencies(root)
    stats = _compute_stats(all_files)

    return Architecture(
        project_name=project_name,
        components=components,
        data_flows=data_flows,
        dependencies=deps,
        file_tree=tree,
        stats=stats,
    )


def _extract_python_imports(fpath: str) -> list[str]:
    """Fast regex-based import extraction."""
    imports = []
    try:
        with open(fpath, "r", encoding="utf-8", errors="ignore") as f:
            for line in f:
                line = line.strip()
                if line.startswith("import "):
                    imports.append(line.split()[1].split(".")[0])
                elif line.startswith("from "):
                    parts = line.split()
                    if len(parts) >= 2:
                        imports.append(parts[1].split(".")[0])
                elif line and not line.startswith("#") and not line.startswith('"""'):
                    # Stop after we pass the import block
                    if not any(line.startswith(kw) for kw in ("import", "from", "#", '"""', "'''", "@")):
                        break
    except Exception:
        pass
    return list(set(imports))


def _insert_tree(tree: dict, parts: list[str], fi: FileInfo):
    """Insert a file into the nested tree dict."""
    node = tree
    for i, part in enumerate(parts):
        if i == len(parts) - 1:
            node[part] = {"_file": fi.model_dump()}
        else:
            if part not in node:
                node[part] = {}
            node = node[part]


def _detect_components(root: Path, files: list[FileInfo]) -> list[Component]:
    """Auto-detect project components from directory structure and files."""
    components: list[Component] = []

    # Backend detection
    for dirname in ("backend", "server", "api", "src"):
        dirpath = root / dirname
        if dirpath.is_dir():
            tech = _detect_backend_tech(dirpath)
            if tech:
                comp_files = [f.path for f in files if f.path.startswith(dirname + "/")]
                routes = _extract_routes(root, comp_files)
                components.append(Component(
                    name=dirname.capitalize(),
                    type=ComponentType.BACKEND,
                    path=dirname,
                    tech_stack=tech,
                    files=comp_files,
                    api_routes=routes,
                    entry_points=_find_entry_points(root / dirname, "backend"),
                ))
                break

    # Check root-level Python project
    if not any(c.type == ComponentType.BACKEND for c in components):
        if (root / "pyproject.toml").exists() or (root / "setup.py").exists():
            app_dir = root / "app"
            if app_dir.is_dir():
                tech = _detect_backend_tech(root)
                comp_files = [f.path for f in files if f.path.startswith("app/")]
                components.append(Component(
                    name="App",
                    type=ComponentType.BACKEND,
                    path=".",
                    tech_stack=tech,
                    files=comp_files,
                    entry_points=_find_entry_points(root, "backend"),
                ))

    # Frontend / Portal detection
    for dirname in ("portal", "frontend", "web", "app", "client", "ui", "dashboard"):
        pkg_path = root / dirname / "package.json"
        if pkg_path.exists():
            tech = _detect_frontend_tech(pkg_path)
            comp_files = [f.path for f in files if f.path.startswith(dirname + "/")]
            components.append(Component(
                name=dirname.capitalize(),
                type=ComponentType.FRONTEND,
                path=dirname,
                tech_stack=tech,
                files=comp_files,
                entry_points=_find_entry_points(root / dirname, "frontend"),
            ))
            break

    # Chrome extension detection
    for dirname in ("chrome-extension", "extension", "ext", "browser-extension"):
        manifest = root / dirname / "manifest.json"
        if manifest.exists():
            tech = ["Chrome MV3"]
            try:
                m = json.loads(manifest.read_text())
                if m.get("manifest_version") == 2:
                    tech = ["Chrome MV2"]
            except Exception:
                pass
            comp_files = [f.path for f in files if f.path.startswith(dirname + "/")]
            components.append(Component(
                name="Extension",
                type=ComponentType.EXTENSION,
                path=dirname,
                tech_stack=tech,
                files=comp_files,
            ))
            break

    # Database detection
    for pattern in ("**/alembic.ini", "**/prisma/schema.prisma", "**/migrations/**"):
        if list(root.glob(pattern)):
            components.append(Component(
                name="Database",
                type=ComponentType.DATABASE,
                path="migrations",
                tech_stack=_detect_db_tech(root),
            ))
            break

    # Docker detection
    dockerfiles = list(root.glob("**/Dockerfile")) + list(root.glob("**/docker-compose*"))
    if dockerfiles:
        comp_files = [str(p.relative_to(root)).replace("\\", "/") for p in dockerfiles]
        components.append(Component(
            name="Docker",
            type=ComponentType.DOCKER,
            path=".",
            tech_stack=["Docker", "Docker Compose"] if any("compose" in str(p) for p in dockerfiles) else ["Docker"],
            files=comp_files,
        ))

    # CI/CD detection
    gh_workflows = root / ".github" / "workflows"
    if gh_workflows.is_dir():
        comp_files = [f.path for f in files if ".github/workflows" in f.path]
        components.append(Component(
            name="CI/CD",
            type=ComponentType.CI_CD,
            path=".github/workflows",
            tech_stack=["GitHub Actions"],
            files=comp_files,
        ))

    # Claude / MCP config
    claude_dir = root / ".claude"
    if claude_dir.is_dir():
        comp_files = [f.path for f in files if f.path.startswith(".claude/")]
        components.append(Component(
            name="Claude Config",
            type=ComponentType.CONFIG,
            path=".claude",
            tech_stack=["Claude Code", "MCP"],
            files=comp_files,
        ))

    return components


def _detect_backend_tech(dirpath: Path) -> list[str]:
    tech = []
    pyproject = dirpath / "pyproject.toml"
    requirements = dirpath / "requirements.txt"

    text = ""
    if pyproject.exists():
        text = pyproject.read_text(errors="ignore")
    elif requirements.exists():
        text = requirements.read_text(errors="ignore")

    if "fastapi" in text.lower():
        tech.append("FastAPI")
    if "django" in text.lower():
        tech.append("Django")
    if "flask" in text.lower():
        tech.append("Flask")
    if "sqlalchemy" in text.lower():
        tech.append("SQLAlchemy")
    if "pydantic" in text.lower():
        tech.append("Pydantic")
    if "celery" in text.lower():
        tech.append("Celery")
    if "redis" in text.lower():
        tech.append("Redis")
    if not tech:
        tech.append("Python")
    return tech


def _detect_frontend_tech(pkg_json: Path) -> list[str]:
    tech = []
    try:
        pkg = json.loads(pkg_json.read_text())
        deps = {**pkg.get("dependencies", {}), **pkg.get("devDependencies", {})}
        if "next" in deps:
            tech.append("Next.js")
        if "react" in deps:
            tech.append("React")
        if "vue" in deps:
            tech.append("Vue")
        if "svelte" in deps or "@sveltejs/kit" in deps:
            tech.append("Svelte")
        if "@angular/core" in deps:
            tech.append("Angular")
        if "tailwindcss" in deps:
            tech.append("Tailwind")
        if "typescript" in deps:
            tech.append("TypeScript")
    except Exception:
        tech.append("Node.js")
    return tech or ["Node.js"]


def _detect_db_tech(root: Path) -> list[str]:
    tech = []
    if (root / "alembic.ini").exists() or list(root.glob("**/alembic/**")):
        tech.append("Alembic")
    if list(root.glob("**/prisma/schema.prisma")):
        tech.append("Prisma")
    # Check for DB URLs in configs
    for cfg in root.glob("**/.env*"):
        try:
            text = cfg.read_text(errors="ignore")
            if "postgres" in text.lower():
                tech.append("PostgreSQL")
            if "sqlite" in text.lower():
                tech.append("SQLite")
            if "mysql" in text.lower():
                tech.append("MySQL")
        except Exception:
            pass
    return tech or ["SQL"]


def _extract_routes(root: Path, files: list[str]) -> list[APIRoute]:
    """Extract API routes from Python FastAPI/Flask files."""
    routes: list[APIRoute] = []
    route_pattern = re.compile(
        r'@(?:app|router)\.(get|post|put|delete|patch|websocket)\s*\(\s*["\']([^"\']+)["\']'
    )

    for fpath in files:
        if not fpath.endswith(".py"):
            continue
        abs_path = root / fpath
        try:
            content = abs_path.read_text(errors="ignore")
            for match in route_pattern.finditer(content):
                method = match.group(1).upper()
                path = match.group(2)
                routes.append(APIRoute(method=method, path=path, file=fpath))
        except Exception:
            continue

    return routes


def _find_entry_points(dirpath: Path, comp_type: str) -> list[str]:
    entries = []
    if comp_type == "backend":
        for name in ("main.py", "app.py", "server.py", "wsgi.py", "asgi.py"):
            for p in dirpath.rglob(name):
                entries.append(str(p.relative_to(dirpath)).replace("\\", "/"))
    elif comp_type == "frontend":
        for name in ("page.tsx", "page.jsx", "index.tsx", "index.jsx", "App.tsx", "App.jsx"):
            for p in dirpath.rglob(name):
                entries.append(str(p.relative_to(dirpath)).replace("\\", "/"))
    return entries[:10]


def _infer_data_flows(components: list[Component], files: list[FileInfo]) -> list[DataFlow]:
    """Infer data flows between components based on imports and known patterns."""
    flows: list[DataFlow] = []
    comp_names = {c.name for c in components}

    # Frontend → Backend (if both exist)
    frontend = next((c for c in components if c.type == ComponentType.FRONTEND), None)
    backend = next((c for c in components if c.type == ComponentType.BACKEND), None)
    if frontend and backend:
        flows.append(DataFlow(
            source=frontend.name,
            target=backend.name,
            protocol="HTTP/REST",
            description="API calls",
            direction=DataFlowDirection.BIDIRECTIONAL,
        ))

    # Extension → Backend (if both exist)
    ext = next((c for c in components if c.type == ComponentType.EXTENSION), None)
    if ext and backend:
        flows.append(DataFlow(
            source=ext.name,
            target=backend.name,
            protocol="WebSocket",
            description="Real-time messaging",
            direction=DataFlowDirection.BIDIRECTIONAL,
        ))

    # Backend → Database
    db = next((c for c in components if c.type == ComponentType.DATABASE), None)
    if backend and db:
        flows.append(DataFlow(
            source=backend.name,
            target=db.name,
            protocol="SQL",
            description="Data persistence",
            direction=DataFlowDirection.BIDIRECTIONAL,
        ))

    return flows


def _collect_dependencies(root: Path) -> list[DependencyInfo]:
    """Collect top-level dependencies from package files."""
    deps: list[DependencyInfo] = []

    # Python
    for pyproject in root.rglob("pyproject.toml"):
        if "node_modules" in str(pyproject):
            continue
        try:
            text = pyproject.read_text(errors="ignore")
            # Simple regex extraction — not a full TOML parser
            in_deps = False
            for line in text.split("\n"):
                if "dependencies" in line and "=" in line and "[" in line:
                    in_deps = True
                    continue
                if in_deps:
                    if line.strip().startswith("]"):
                        in_deps = False
                        continue
                    match = re.match(r'\s*"([^">=<\[]+)', line)
                    if match:
                        deps.append(DependencyInfo(name=match.group(1).strip(), category="python"))
        except Exception:
            pass

    # Node
    for pkg_json in root.rglob("package.json"):
        if "node_modules" in str(pkg_json):
            continue
        try:
            pkg = json.loads(pkg_json.read_text())
            for name in pkg.get("dependencies", {}):
                deps.append(DependencyInfo(
                    name=name,
                    version=pkg["dependencies"][name],
                    category="node-runtime",
                ))
            for name in pkg.get("devDependencies", {}):
                deps.append(DependencyInfo(
                    name=name,
                    version=pkg["devDependencies"][name],
                    category="node-dev",
                ))
        except Exception:
            pass

    return deps


def _compute_stats(files: list[FileInfo]) -> dict:
    by_cat: dict[str, int] = {}
    by_lang: dict[str, int] = {}
    total_size = 0

    for f in files:
        by_cat[f.category] = by_cat.get(f.category, 0) + 1
        by_lang[f.language] = by_lang.get(f.language, 0) + 1
        total_size += f.size

    return {
        "total_files": len(files),
        "total_size": total_size,
        "by_category": by_cat,
        "by_language": by_lang,
    }
